### ISSUE TITLE
Supervised Topic Modelling unable to produce same output 

### ISSUE THREAD

**Issue by rrachelsee** [MaartenGr/BERTopic#1270]
Hi, 
I referenced the code from https://maartengr.github.io/BERTopic/getting_started/supervised/supervised.html for BERTopic Supervised modelling. Although I was able to obtain the correct output using fit_transform(docs, y=y), I was unable to obtain the same output using transform(docs) even though the inputted doc is the same. Most of the topics generated were labelled -1. 

May I know what has gone wrong? 

For reference, this is my model code: 
empty_dimensionality_model = BaseDimensionalityReduction()
clf = LogisticRegression()
ctfidf_model = ClassTfidfTransformer(reduce_frequent_words=True)

topic_model = BERTopic(
        umap_model=empty_dimensionality_model,
        hdbscan_model=clf,
        ctfidf_model=ctfidf_model,
        n_gram_range=(1,3))

Thank you! Any help would be truly appreciated.

**Comment by MaartenGr** [comment]
Could you create a reproducible example? Without it is difficult to see what exactly is going wrong. Also, it might help to set a `random_state` when using LogisticRegression.

**Comment by phorvicheka** [comment]
Hello,

I got the same problem.

```
from bertopic import BERTopic
from bertopic.vectorizers import ClassTfidfTransformer
from bertopic.dimensionality import BaseDimensionalityReduction
from sklearn.linear_model import LogisticRegression
from sklearn.datasets import fetch_20newsgroups
from sentence_transformers import SentenceTransformer
from umap import UMAP


################## Data Preparation ##################
# Get labeled data
data = fetch_20newsgroups(subset="all", remove=("headers", "footers", "quotes"))
docs = data["data"]
y = data["target"]

################## Topic modelling ##################

##Prepare embeddings
sentence_model = SentenceTransformer("all-mpnet-base-v2")
embeddings = sentence_model.encode(docs, show_progress_bar=True)

##Train model
empty_dimensionality_model = BaseDimensionalityReduction()
clf = LogisticRegression(random_state=42)
ctfidf_model = ClassTfidfTransformer(reduce_frequent_words=True)

base_topic_model = BERTopic(
    umap_model=empty_dimensionality_model,
    hdbscan_model=clf,
    ctfidf_model=ctfidf_model,
    verbose=True,
    # calculate_probabilities=True,
)
topics, probs = base_topic_model.fit_transform(docs, embeddings, y=y)

##Transform same documents and check if the predicted topics are the same
new_topics, new_probs = base_topic_model.transform(docs, embeddings)
assert topics == new_topics
```

![2023-09-09_13h30_09](https://github.com/MaartenGr/BERTopic/assets/16742458/75e4a261-ab73-4439-a178-012f9e19be6b)

**Comment by MaartenGr** [comment]
@phorvicheka I believe what is happening your case is the following. When you run `.fit_transform` with a prediction model, it returns the `y` variable instead of the actual predictions. Since the model should be fitted in a supervised task and you already assigned documents to clusters, the `y` variable is directly used to create as "predictions". Although they are not technically predictions. 

Then, when you run `.transform`, the underlying prediction is indeed used to run `.predict` in order to generate the prediction since we assume that `.transform` is used to generated predictions for unseen documents. 

To sum up, in `.fit_transform` of BERTopic only `.fit` of your LogisticRegression is used and `y` is returned without actually running the prediction. In contrast, in `.transform` of BERTopic, the `.predict` of your LogisticRegression is used and its predictions are returned. If both topics would be the same, that would actually mean your model is overfitting.

**Comment by phorvicheka** [comment]
<blockquote>If both topics would be the same, that would actually mean your model is overfitting.</blockquote>

@MaartenGr Thank you very much for quick response. I don't understand well since I am newbie for this field. What I understand is that the output of topics from `fit_transform(docs, y=y)` and `transform(docs)`, are not supposed to be the same for *Supervised Topic Modelling*.

**Comment by MaartenGr** [comment]
@phorvicheka Yes, but it is also important to understand why. It means that when you train a classifier, it seldom happens that when you train on `y` it will perfectly recreate `y` between `.fit` and `.transform`. A classification algorithm will nearly always be a proxy of the original `y` and will try to recreate it. No classification algorithm is 100% accurate, and that is also what you see here.
